src/Dependencies/PooledObjects/ArrayBuilder.cs

```diff
@@ -726,6 +744,194 @@ public ImmutableArray<S> SelectDistinct<S>(Func<T, S> selector)
             return result.ToImmutableAndFree();
         }
 
+        public bool Any(Func<T, bool> predicate)
+        {
+            foreach (var item in this)
+            {
+                if (predicate(item))
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        public bool Any<A>(Func<T, A, bool> predicate, A arg)
+        {
+            foreach (var item in this)
+            {
+                if (predicate(item, arg))
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        public bool All(Func<T, bool> predicate)
+        {
+            foreach (var item in this)
+            {
+                if (!predicate(item))
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        public bool All<A>(Func<T, A, bool> predicate, A arg)
+        {
+            foreach (var item in this)
+            {
+                if (!predicate(item, arg))
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Maps an array builder to immutable array.
+        /// </summary>
+        /// <typeparam name="TResult"></typeparam>
+        /// <param name="map">The mapping delegate</param>
+        /// <returns>If the items's length is 0, this will return an empty immutable array</returns>
+        public ImmutableArray<TResult> SelectAsArray<TResult>(Func<T, TResult> map)
+        {
+            switch (Count)
+            {
+                case 0:
+                    return [];
+
+                case 1:
+                    return [map(this[0])];
+
+                case 2:
+                    return [map(this[0]), map(this[1])];
+
+                case 3:
+                    return [map(this[0]), map(this[1]), map(this[2])];
+
+                case 4:
+                    return [map(this[0]), map(this[1]), map(this[2]), map(this[3])];
+
+                default:
+                    var builder = ArrayBuilder<TResult>.GetInstance(Count);
```

Comment:
>var builder = ArrayBuilder<TResult>.GetInstance(Count); [](http://example.com/codeflow?start=20&length=55)

Looks like this is just a direct port, but it seems like this would be better if it used ImmutableCollectionsMarshal with a newly created array rather than churning the pool.

Reply:
FixedSizeArrayBuilder uses that. Another advantage would be that this could be changed to only special case the 0-sized array.

Reply:
I just moved the existing impl + used collection exprs. Feel free to send follow up PR with improvements

Reply:
Yup. Let's just move, and do the bare minimum in terms of changes. Follow-ups definitely welcome